### PR TITLE
Loosened rules involving statics mentioning other statics

### PR DESCRIPTION
**Pull request by alexreg**
Before this PR, trying to mention a static in any way other than taking a reference to it caused a compile-time error. So, while

```rust
static A: u32 = 42;
static B: &u32 = &A;
```

compiles successfully,

```rust
static A: u32 = 42;
static B: u32 = A; // error
```

and 

```rust
static A: u32 = 42;
static B: u32 = *&A; // error
```

are not possible to express in Rust. On the other hand, introducing an intermediate `const fn` can presently allow one to do just that:

```rust
static A: u32 = 42;
static B: u32 = foo(&A); // success!

const fn foo(a: &u32) -> u32 {
    *a
}
```

Preventing `const fn` from allowing to work around the ban on reading from statics would cripple `const fn` almost into uselessness.
Additionally, the limitation for reading from statics comes from the old const evaluator(s) and is not shared by `miri`.

This PR loosens the rules around use of statics to allow statics to evaluate other statics by value, allowing all of the above examples to compile and run successfully.
Reads from extern (foreign) statics are however still disallowed by miri, because there is no compile-time value to be read.

```rust
extern static A: u32;

static B: u32 = A; // error
```

This opens up a new avenue of potential issues, as a static can now not just refer to other statics or read from other statics, but even contain references that point into itself.
While it might seem like this could cause subtle bugs like allowing a static to be initialized by its own value, this is inherently impossible in miri.
Reading from a static causes the `const_eval` query for that static to be invoked. Calling the `const_eval` query for a static while already inside the `const_eval` query of said static will cause cycle errors.
It is not possible to accidentally create a bug in miri that would enable initializing a static with itself, because the memory of the static *does not exist* while being initialized.
The memory is not uninitialized, it is not there. Thus any change that would accidentally allow reading from a not yet initialized static would cause ICEs.

Tests have been modified according to the new rules, and new tests have been added for writing to `static mut`s within definitions of statics (which needs to fail), and incremental compilation with complex/interlinking static definitions.
Note that incremental compilation did not need to be adjusted, because all of this was already possible before with workarounds (like intermediate `const fn`s) and the encoding/decoding already supports all the possible cases.

r? @eddyb

### CODE REVIEW

File: src/librustc_mir/transform/qualify_consts.rs [hunk 1]
```diff
@@ -228,38 +222,7 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
 
     /// Check if a Local with the current qualifications is promotable.
     fn can_promote(&self, qualif: Qualif) -> bool {
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Can you inline `can_promote` now?

File: src/librustc_mir/transform/qualify_consts.rs [hunk 2]
```diff
@@ -492,7 +455,9 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
         match *place {
             Place::Local(ref local) => self.visit_local(local, context, location),
             Place::Static(ref global) => {
-                self.add(Qualif::STATIC);
+                if !(self.mode == Mode::Static || self.mode == Mode::StaticMut) {
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Can you add a comment above this `if` saying that only statics can refer to other statics?

File: src/librustc_mir/transform/qualify_consts.rs [hunk 3]
```diff
@@ -677,16 +618,11 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
             Rvalue::Len(_) => {
                 // Static places in consts would have errored already,
                 // don't treat length checks as reads from statics.
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Also remove the comment here.

**Comment by eddyb** [comment]
cc @oli-obk @nikomatsakis This PR allows `static`s to read from other statics (although `miri` may still error as before, when the operation is actually attempted), which was technically already possible:
```rust
const fn read<T: Copy>(r: &T) -> T { *r }
static FOO: u32 = read(&BAR);
static BAR: u32: 10;
```
That is, the analysis being removed here was an unsound approximation.
However, `const` and `const fn` still cannot *even mention* a `static`, neither for borrowing or reading.

In terms of motivation, simplifying the analysis `qualify_consts` will allow us to move it to the dataflow framework to support arbitrary CFG shapes, not just linear control-flow.
Does this PR require going through FCP or even an RFC?

**Comment by alexreg** [comment]
@eddyb Okay, addressed your comments and removed the test that's not longer applicable. Should be passing now.

Here's to hoping this PR can either be signed off straight away by a team member or go through an FCP at worst.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 4]
```diff
@@ -226,42 +220,6 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
...
-    fn try_consume(&mut self) -> bool {
-        if self.qualif.intersects(Qualif::STATIC) && self.mode != Mode::Fn {
-            let msg = if self.mode == Mode::Static ||
-                         self.mode == Mode::StaticMut {
-                "cannot refer to other statics by value, use the \
-                 address-of operator or a constant instead"
-            } else {
-                "cannot refer to statics by value, use a constant instead"
-            };
-            struct_span_err!(self.tcx.sess, self.span, E0394, "{}", msg)
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Oh you also need to remove the diagnostic definition of `E0394` (just grep for that error code).

Comment (alexreg) [reply src/librustc_mir/transform/qualify_consts.rs]: And E0494 too, I presume?

File: src/librustc_mir/transform/qualify_consts.rs [hunk 5]
```diff
@@ -566,12 +518,6 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
 
                         ProjectionElem::Field(..) |
                         ProjectionElem::Index(_) => {
-                            if this.mode != Mode::Fn &&
-                               this.qualif.intersects(Qualif::STATIC) {
-                                span_err!(this.tcx.sess, this.span, E0494,
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Same with this one too.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 6]
```diff
@@ -305,7 +263,7 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
             }) if self.mir.local_kind(index) == LocalKind::Temp
                && self.mir.local_decls[index].ty.is_box()
                && self.local_qualif[index].map_or(false, |qualif| {
-                    qualif.intersects(Qualif::NOT_CONST)
+                    qualif.contains(Qualif::NOT_CONST)
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Why did you change `intersects` to `contains`?

**Comment by alexreg** [comment]
Okay, got it. Sorry.

**Comment by alexreg** [comment]
@eddyb What should we do about the above? In particular the "error: compiler encountered internal error".

File: src/librustc_mir/transform/qualify_consts.rs [hunk 7]
```diff
@@ -492,17 +450,31 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
...
                 if self.mode != Mode::Fn {
-                    for attr in &self.tcx.get_attrs(global.def_id)[..] {
-                        if attr.check_name("thread_local") {
-                            span_err!(self.tcx.sess, self.span, E0625,
-                                      "thread-local statics cannot be \
-                                       accessed at compile-time");
-                            self.add(Qualif::NOT_CONST);
-                            return;
-                        }
+                    let static_kind = if self.tcx.is_foreign_item(global.def_id) {
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: s/`static_kind`/`runtime_static`

File: src/test/compile-fail/check-static-recursion-foreign.rs [hunk 8]
```diff
@@ -29,5 +29,6 @@ extern "C" {
 }
 
 static B: &'static c_int = unsafe { &test_static };
+//~^ ERROR foreign statics cannot be accessed at compile-time
```
Comment (eddyb) [inline src/test/compile-fail/check-static-recursion-foreign.rs]: Wait, this is correct... I think we have to allow it, and instead make `miri` disallow *reads* from them. Sorry about the detour, I didn't realize borrowing `extern static` was already allowed.

File: src/test/compile-fail/issue-17718-borrow-interior.rs [hunk 9]
```diff
@@ -1,25 +0,0 @@
...
-//~^ ERROR: cannot refer to the interior of another static
-
-static D: [usize; 1] = [1];
-static E: usize = D[0];
-//~^ ERROR: cannot refer to the interior of another static
-//~^^ ERROR: cannot refer to other statics by value
-static F: &'static usize = &D[0];
-//~^ ERROR: cannot refer to the interior of another static
-
-fn main() {}
```
Comment (eddyb) [inline src/test/compile-fail/issue-17718-borrow-interior.rs]: Can you move this test to `run-pass`?

Comment (alexreg) [reply src/test/compile-fail/issue-17718-borrow-interior.rs]: So, do you want to leave this and the above change to miri to another PR, or do this here?

**Comment by alexreg** [comment]
@eddyb Okay, I think I've solved the MIRI issue with foreign statics. See the test failures above... do I only need to update the comments in the compile-fail tests now, and we're all good?

File: src/test/run-pass/issue-17718-borrow-interior.rs [hunk 10]
```diff
@@ -9,17 +9,13 @@
...
-//~^ ERROR: cannot refer to the interior of another static
 
 static D: [usize; 1] = [1];
 static E: usize = D[0];
-//~^ ERROR: cannot refer to the interior of another static
-//~^^ ERROR: cannot refer to other statics by value
 static F: &'static usize = &D[0];
-//~^ ERROR: cannot refer to the interior of another static
 
 fn main() {}
```
Comment (oli-obk) [inline src/test/run-pass/issue-17718-borrow-interior.rs]: I think it would be good to add some runtime tests here that assert all the statics' values in order to make sure that the right fields have been read

**Review by oli-obk — COMMENTED** [review]
You accidentally committed a `BIN` file.

Yes, please update the compile-fail test error messages.

File: src/librustc_mir/transform/qualify_consts.rs (hunk 7 above)
Comment (eddyb) [reply src/librustc_mir/transform/qualify_consts.rs]: This is irrelevant anyway now since we can't actually deny anything here.

File: src/librustc/mir/interpret/error.rs [hunk 11]
```diff
@@ -113,6 +113,7 @@ pub enum EvalErrorKind<'tcx, O> {
     Layout(layout::LayoutError<'tcx>),
     HeapAllocZeroBytes,
     HeapAllocNonPowerOfTwoAlignment(u64),
+    ForeignStatic,
```
Comment (eddyb) [inline src/librustc/mir/interpret/error.rs]: This should probably be `ReadFromForeignStatic`.

**Comment by eddyb** [comment]
@alexreg You have a binary file named `63` in one of the commits, that you need to remove.

**Comment by nikomatsakis** [comment]
@eddyb 

> In terms of motivation, simplifying the analysis qualify_consts performs will allow us to move it to the dataflow framework to support arbitrary CFG shapes, not just linear control-flow.
Does this PR require going through FCP or even an RFC?

My take: I'm not sure what the best process is here per se, but I think what I really want is that we have these rules documented in the Rust reference (or some other source). In that case, I think that potentially FCP'ing a minor alteration (without a full blown RFC) wouldn't be so bad. It's pretty in the weeds. But an RFC also wouldn't be bad.

I feel like we need to revamp our process slightly in this area in general, to try and move towards greater clarity on what the intended semantics are for various things (and — specifically — where those semantics ought to be specified).

**Comment by alexreg** [comment]
@eddyb 

> @alexreg You have a binary file named 63 in one of the commits, that you need to remove.
> Also, can you squash your commits into one? Or two, one for qualify_consts and one for miri.

Sure. Is this the `BIN` fie @oli-obk was referring to?

File: src/librustc/mir/interpret/error.rs (hunk 11 above)
Comment (eddyb) [reply src/librustc/mir/interpret/error.rs]: No, that's why we can't error in const_qualif - borrowing a foreign static is perfectly fine.

Comment (alexreg) [reply src/librustc/mir/interpret/error.rs]: Oh yes, duh.

**Comment by alexreg** [comment]
Here's the miri error with backtrace for the remaining ICE. (Everything else is passing.)

```
---- [compile-fail] compile-fail/static-array-across-crate.rs stdout ----

error: compiler panicked
status: exit code: 101
command: "/Users/alex/Software/rust-devel/build/x86_64-apple-darwin/stage2/bin/rustc" "/Users/alex/Software/rust-devel/src/test/compile-fail/static-array-across-crate.rs" "--target=x86_64-apple-darwin" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "-o" "/Users/alex/Software/rust-devel/build/x86_64-apple-darwin/test/compile-fail/static-array-across-crate/a" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/Users/alex/Software/rust-devel/build/x86_64-apple-darwin/native/rust-test-helpers" "-L" "/Users/alex/Software/rust-devel/build/x86_64-apple-darwin/test/compile-fail/static-array-across-crate/auxiliary" "-A" "unused"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
thread 'main' panicked at 'already borrowed: BorrowMutError', libcore/result.rs:945:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: rustc::util::common::panic_hook
   5: std::panicking::rust_panic_with_hook
   6: std::panicking::begin_panic_fmt
   7: rust_begin_unwind
   8: core::panicking::panic_fmt
   9: core::result::unwrap_failed
  10: <rustc_mir::interpret::memory::Memory<'a, 'mir, 'tcx, M>>::allocate_value
  11: rustc_mir::interpret::const_eval::eval_body_using_ecx
  12: rustc_mir::interpret::const_eval::eval_body_and_ecx
  13: rustc_mir::interpret::const_eval::const_eval_provider
  14: rustc::ty::maps::__query_compute::const_eval
  15: rustc::ty::maps::<impl rustc::ty::maps::config::QueryConfig<'tcx> for rustc::ty::maps::queries::const_eval<'tcx>>::compute
  16: rustc::dep_graph::graph::DepGraph::with_task_impl
  17: rustc::ty::context::tls::with_related_context
  18: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  19: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  20: rustc::ty::maps::TyCtxtAt::const_eval
  21: <rustc_mir::interpret::memory::Memory<'a, 'mir, 'tcx, M>>::get
  22: <rustc_mir::interpret::memory::Memory<'a, 'mir, 'tcx, M>>::check_align
  23: <rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::try_read_value
  24: <rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::write_value
  25: <rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::eval_rvalue_into_place
  26: rustc_mir::interpret::step::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::step
  27: rustc_mir::interpret::const_eval::eval_body_using_ecx
  28: rustc_mir::interpret::const_eval::eval_body_and_ecx
  29: rustc_mir::interpret::const_eval::const_eval_provider
  30: rustc::ty::maps::__query_compute::const_eval
  31: rustc::ty::maps::<impl rustc::ty::maps::config::QueryConfig<'tcx> for rustc::ty::maps::queries::const_eval<'tcx>>::compute
  32: rustc::dep_graph::graph::DepGraph::with_task_impl
  33: rustc::ty::context::tls::with_related_context
  34: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  35: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  36: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::const_eval
  37: rustc_mir::monomorphize::collector::collect_items_rec
  38: rustc_mir::monomorphize::collector::collect_crate_mono_items
  39: rustc::util::common::time
  40: rustc_codegen_llvm::base::collect_and_partition_mono_items
  41: rustc::ty::maps::<impl rustc::ty::maps::config::QueryConfig<'tcx> for rustc::ty::maps::queries::collect_and_partition_mono_items<'tcx>>::compute
  42: rustc::dep_graph::graph::DepGraph::with_task_impl
  43: rustc::ty::context::tls::with_related_context
  44: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  45: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  46: rustc_codegen_llvm::base::codegen_crate
  47: <rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_utils::codegen_backend::CodegenBackend>::codegen_crate
  48: rustc::util::common::time
  49: rustc_driver::driver::phase_4_codegen
  50: rustc_driver::driver::compile_input::{{closure}}
  51: rustc::ty::context::tls::enter_context
  52: <std::thread::local::LocalKey<T>>::with
  53: rustc::ty::context::TyCtxt::create_and_enter
  54: rustc_driver::driver::compile_input
  55: rustc_driver::run_compiler_with_pool
  56: syntax::with_globals
  57: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  58: __rust_maybe_catch_panic
  59: std::panicking::try
  60: rustc_driver::run
  61: rustc_driver::main
  62: std::rt::lang_start::{{closure}}
  63: std::panicking::try::do_call
  64: __rust_maybe_catch_panic
  65: std::panic::catch_unwind
  66: std::rt::lang_start_internal
  67: main
query stack during panic:
{"message":"#0 [const_eval] const-evaluating `array::ARRAY`","code":null,"level":"","spans":[{"file_name":"/Users/alex/Software/rust-devel/src/test/compile-fail/static-array-across-crate.rs","byte_start":801,"byte_end":812,"line_start":23,"line_end":23,"column_start":16,"column_end":27,"is_primary":true,"text":[{"text":"static Z: u8 = (&ARRAY)[0];","highlight_start":16,"highlight_end":27}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"#0 [const_eval] const-evaluating `array::ARRAY`\n  --> /Users/alex/Software/rust-devel/src/test/compile-fail/static-array-across-crate.rs:23:16\n   |\nLL | static Z: u8 = (&ARRAY)[0];\n   |                ^^^^^^^^^^^\n"}
{"message":"#1 [const_eval] const-evaluating `Z`","code":null,"level":"","spans":[{"file_name":"/Users/alex/Software/rust-devel/src/test/compile-fail/static-array-across-crate.rs","byte_start":0,"byte_end":0,"line_start":1,"line_end":1,"column_start":1,"column_end":1,"is_primary":true,"text":[{"text":"// Copyright 2015 The Rust Project Developers. See the COPYRIGHT","highlight_start":1,"highlight_end":1}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"#1 [const_eval] const-evaluating `Z`\n"}
{"message":"#2 [collect_and_partition_mono_items] collect_and_partition_mono_items","code":null,"level":"","spans":[{"file_name":"/Users/alex/Software/rust-devel/src/test/compile-fail/static-array-across-crate.rs","byte_start":0,"byte_end":0,"line_start":1,"line_end":1,"column_start":1,"column_end":1,"is_primary":true,"text":[{"text":"// Copyright 2015 The Rust Project Developers. See the COPYRIGHT","highlight_start":1,"highlight_end":1}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"#2 [collect_and_partition_mono_items] collect_and_partition_mono_items\n"}
end of query stack

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.28.0-dev running on x86_64-apple-darwin

note: compiler flags: -Z ui-testing -Z unstable-options -C prefer-dynamic -C rpath
```

File: src/test/compile-fail/issue-28324.rs [hunk 12]
```diff
@@ -15,6 +15,6 @@ extern {
 }
 
 pub static BAZ: u32 = *&error_message_count;
-//~^ ERROR cannot refer to other statics by value
+//~^ ERROR constant evaluation error
```
Comment (eddyb) [inline src/test/compile-fail/issue-28324.rs]: Can you add another line with `//~| read from foreign (extern) static` just to make it clear what error it triggers?

Comment (alexreg) [reply src/test/compile-fail/issue-28324.rs]: @eddyb That gives a "not found" error unfortunately.

Comment (alexreg) [reply src/test/compile-fail/issue-28324.rs]: ...since I don't think it's actually part of the error message.

Comment (eddyb) [reply src/test/compile-fail/issue-28324.rs]: Wait, what? What do you get if you compile that test locally, directly?

**Comment by oli-obk** [comment]
Can you add a test that tries to write to a static mut in the initializer of another static while the `const_let` feature gate is active?

And a test that does the same, but tries to write to the static itself.

On that note, a test where a static tries to initialize itself with its own value would probably be a good idea, too

**Comment by alexreg** [comment]
@eddyb @oli-obk Okay, I think we're ready to FCP this. :-)

**Comment by alexreg** [comment]
@oli-obk @eddyb Given there was not time to discuss this in the meeting, does one of you want to nominate it for FCP so team members can raise their concerns (if any) here?

**Comment by nikomatsakis** [comment]
@alexreg Hey! Sorry for being slow to respond to your pings. @oli-obk has promised me a write-up that summarizes the effects of this change -- once they've finished that, I will happily instruct rfcbot to propose FCP. Also, they're here in person and can bother me to ensure this happens. =)

**Comment by alexreg** [comment]
@nikomatsakis No problem! I probably should have written something up myself, but since @oli-obk understands this stuff a bit better, that sounds fine to me. Thanks.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

I propose that we adopt this tweak to the rules. The description by @alexreg of the changes follows:

---

Before this PR, trying to mention a static in any way other than taking a reference to it caused a compile-time error. So, while

```rust
static A: u32 = 42;
static B: &u32 = &A;
```

compiles successfully,

```rust
static A: u32 = 42;
static B: u32 = A; // error
```

and 

```rust
static A: u32 = 42;
static B: u32 = *&A; // error
```

are not possible to express in Rust. On the other hand, introducing an intermediate `const fn` can presently allow one to do just that:

```rust
static A: u32 = 42;
static B: u32 = foo(&A); // success!

const fn foo(a: &u32) -> u32 {
    *a
}
```

Preventing `const fn` from allowing to work around the ban on reading from statics would cripple `const fn` almost into uselessness.
Additionally, the limitation for reading from statics comes from the old const evaluator(s) and is not shared by `miri`.

This PR loosens the rules around use of statics to allow statics to evaluate other statics by value, allowing all of the above examples to compile and run successfully.
Reads from extern (foreign) statics are however still disallowed by miri, because there is no compile-time value to be read.

```rust
extern static A: u32;

static B: u32 = A; // error
```

This opens up a new avenue of potential issues, as a static can now not just refer to other statics or read from other statics, but even contain references that point into itself.
While it might seem like this could cause subtle bugs like allowing a static to be initialized by its own value, this is inherently impossible in miri.
Reading from a static causes the `const_eval` query for that static to be invoked. Calling the `const_eval` query for a static while already inside the `const_eval` query of said static will cause cycle errors.
It is not possible to accidentally create a bug in miri that would enable initializing a static with itself, because the memory of the static *does not exist* while being initialized.
The memory is not uninitialized, it is not there. Thus any change that would accidentally allow reading from a not yet initialized static would cause ICEs.

Tests have been modified according to the new rules, and new tests have been added for writing to `static mut`s within definitions of statics (which needs to fail), and incremental compilation with complex/interlinking static definitions.
Note that incremental compilation did not need to be adjusted, because all of this was already possible before with workarounds (like intermediate `const fn`s) and the encoding/decoding already supports all the possible cases.

**Comment by alexreg** [comment]
@nikomatsakis Thanks for kicking this off!

**Comment by alexreg** [comment]
Hmm, shouldn't FCP begin now, or does it take time for @rfcbot to notice?

**Comment by scottmcm** [comment]
@alexreg It's usually picked up immediately; I think more checkboxes are needed before FCP.

**Comment by alexreg** [comment]
@scottmcm Oh. In that case, might be good to change @rfcbot's message, since we already have a majority. :-)

**Comment by scottmcm** [comment]
@alexreg There's some nuance that I don't remember, like it might actually be "go to FCP once everyone but two have signed off, so long as that's a majority".

File: src/librustc_mir/interpret/memory.rs [hunk 13]
```diff
@@ -313,7 +316,8 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'a, 'mir, 'tcx, M> {
                 Some(alloc) => Ok(alloc),
                 None => {
                     // static alloc?
-                    match self.tcx.alloc_map.lock().get(id) {
+                    let alloc = self.tcx.alloc_map.lock().get(id);
+                    match alloc {
```
Comment (bjorn3) [inline src/librustc_mir/interpret/memory.rs]: Why this change?

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: src/librustc_mir/interpret/memory.rs (hunk 13 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/memory.rs]: because otherwise the `alloc_map` stays locked within the match arms, which would not allow the `Static` case to access it again.

Comment (bjorn3) [reply src/librustc_mir/interpret/memory.rs]: I get it

**Comment by alexreg** [comment]
@oli-obk / @eddyb Merge tim? :-D

File: src/librustc/mir/interpret/error.rs [hunk 14]
```diff
@@ -166,6 +167,8 @@ impl<'tcx, O> EvalErrorKind<'tcx, O> {
                 "a raw memory access tried to access part of a pointer value as raw bytes",
             ReadBytesAsPointer =>
                 "a memory access tried to interpret some bytes as a pointer",
+            ReadForeignStatic =>
+                "tried to read foreign (extern) static",
```
Comment (eddyb) [inline src/librustc/mir/interpret/error.rs]: I think this should say "... read *from* foreign ...", not just "read".

File: src/librustc_mir/transform/qualify_consts.rs [hunk 15]
```diff
@@ -492,17 +450,21 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
         match *place {
             Place::Local(ref local) => self.visit_local(local, context, location),
             Place::Static(ref global) => {
-                self.add(Qualif::STATIC);
+                // Only allow statics (not consts) to refer to other statics.
+                if !(self.mode == Mode::Static || self.mode == Mode::StaticMut) {
+                    self.add(Qualif::NOT_CONST);
+                }
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: I just noticed this, but if you move this below the `thread_local` check, you can do this:
```rust
                // Only allow statics (not consts) to refer to other statics.
                if self.mode == Mode::Static || self.mode == Mode::StaticMut {
                    return;
                }
                self.add(Qualif::NOT_CONST);
                if self.mode != Mode::Fn {
                    let mut err = struct_span_err!(self.tcx.sess, self.span, E0013,
```
That is, you can rearrange the code such that error reporting is predicated on `self.mode != Mode::Fn`.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 16]
```diff
@@ -492,17 +450,21 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
...
-                        }
+                    if self.tcx
+                           .get_attrs(global.def_id)
+                           .iter()
+                           .any(|attr| attr.check_name("thread_local")) {
+                        span_err!(self.tcx.sess, self.span, E0625,
+                                    "thread-local statics cannot be \
+                                    accessed at compile-time");
+                        self.add(Qualif::NOT_CONST);
+                        return;
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: You can also changed this so the attribute is *always* checked, and `self.add(Qualif::NOT_CONST); return;` always happen, only the error reporting stays conditional on `if self.mode != Mode::Fn`.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 17]
```diff
@@ -599,7 +549,6 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
             Operand::Move(_) => {
                 self.nest(|this| {
                     this.super_operand(operand, location);
-                    this.try_consume();
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: You can remove the `self.nest` around the `super_operand` call - its only purpose was to observe *only* the flags being added by `super_operand`, in `try_consume`.

Comment (eddyb) [reply src/librustc_mir/transform/qualify_consts.rs]: In fact, you can put `self.super_operand(operand, location)` at the top of the function, before the `match`.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 18]
```diff
@@ -651,14 +600,6 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
...
                         kind
                     }, location);
-                    if !this.try_consume() {
-                        return;
-                    }
-
-                    if this.qualif.intersects(Qualif::STATIC_REF) {
-                        this.qualif = this.qualif - Qualif::STATIC_REF;
-                        this.add(Qualif::STATIC);
-                    }
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Similar thing here, doesn't need the `nest` anymore.

File: src/librustc_mir/transform/qualify_consts.rs (hunk 16 above)
Comment (alexreg) [reply src/librustc_mir/transform/qualify_consts.rs]: Makes sense.

File: src/librustc_mir/transform/qualify_consts.rs (hunk 17 above)
Comment (alexreg) [reply src/librustc_mir/transform/qualify_consts.rs]: Got it.

**Comment by alexreg** [comment]
@eddyb What caused the latest Travis failures? Did something change in master when I rebased? (Because everything was passing before the clean-up, i.e. last commit.) I'm wary of simply hanging the expected error messages.

**Comment by oli-obk** [comment]
@alexreg due to conflicts you need to rebase anyway. Rerun the test suite after you do, it looks like master changed indeed. I did recently change something around error reporting, so I think you can just adjust the error message comments to fit reality

**Comment by alexreg** [comment]
@oli-obk Okay, will do.

File: src/test/compile-fail/issue-14227.rs [hunk 19]
```diff
@@ -13,6 +13,10 @@
 extern {
     pub static symbol: ();
 }
-static CRASH: () = symbol; //~ cannot refer to other statics by value
+static CRASH: () = symbol;
+//~^ ERROR could not evaluate static initializer
```
Comment (oli-obk) [inline src/test/compile-fail/issue-14227.rs]: open an issue about it. this feels like a regression by my diagnostics changes
